src/model_theory/rings/language.lean

```diff
@@ -0,0 +1,76 @@
+/-
+Copyright (c) 2022 Joseph Hua. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Hua
+-/
+import model_theory.terms_and_formulas
+
+/-!
+# The language of rings
+
+## Main Definitions
+* A `first_order.language.ring.L` defines the language of rings,
+  which consists of `0`, `1`, `-`, `+`, `*`.
+-/
+universes u v u'
+
+namespace first_order
+namespace language
+namespace ring
+
+/-- The constant symbols in `first_order.language.ring.L` -/
+inductive consts : Type u
+| zero : consts
+| one : consts
+
+/-- The unary function symbols in `first_order.language.ring.L` -/
+inductive unaries : Type u
+| neg : unaries
+
+/-- The binary function symbols in `first_order.language.ring.L` -/
+inductive binaries : Type u
+| add : binaries
+| mul : binaries
+
+/-- All function symbols in `first_order.language.ring.L` -/
+def functions : ℕ → Type u
+| 0 := consts
+| 1 := unaries
+| 2 := binaries
+| (n + 3) := pempty
+
+instance : inhabited consts := ⟨ consts.zero ⟩
+instance : inhabited unaries := ⟨ unaries.neg ⟩
+instance : inhabited binaries := ⟨ binaries.add ⟩
+instance : inhabited (functions 0) := ⟨ consts.zero ⟩
+
+/-- The language of rings -/
+@[simps] def L : language.{u v} :=
+{ functions := functions,
+  relations := λ n, pempty }
+
+instance : is_algebraic L := { empty_relations :=
+λ n, by {simp, apply_instance} }
```

Comment:
Style:
```suggestion
instance : is_algebraic L :=
{ empty_relations := λ n, by { simp, apply_instance } }
```
Also, this is a nonterminal simp - can you squeeze it?

Reply:
> > > Can you add the `is_algebraic` instance, or does lean already know how to find it?
> > 
> > 
> > I need to add it since it needs to reduce the definition first.
> 
> It _might_ be reasonable to mark the definition `reducible`, in which case you wouldn't need to.

what do you recommend?